### ISSUE TITLE
Cannot read property 'collections' of undefined running unit tests.

### ISSUE THREAD

**Issue by ronsuez** [balderdashy/sails-mongo#278]
I'm running an unit test suite for sails, but when i try to run it, i get this error: 

```javascript
     var collection = connectionObject.collections[collectionName];
                                       ^
TypeError: Cannot read property 'collections' of undefined
    at Object.module.exports.adapter.create (/Users/ron/Projects/sails/ap/node_modules/sails-mongo/lib/adapter.js:268:40)
    at module.exports.create (/Users/ron/Projects/sails/ap/node_modules/sails/node_modules/waterline/lib/waterline/adapter/dql.js:85:13)
    at bound.createValues (/Users/ron/Projects/sails/ap/node_modules/sails/node_modules/waterline/lib/waterline/query/dql/create.js:213:16)
    at /Users/ron/Projects/sails/ap/node_modules/sails/node_modules/waterline/lib/waterline/query/dql/create.js:74:20
    at /Users/ron/Projects/sails/ap/node_modules/async/lib/async.js:254:17
    at /Users/ron/Projects/sails/ap/node_modules/async/lib/async.js:157:25
    at /Users/ron/Projects/sails/ap/node_modules/async/lib/async.js:251:21
    at /Users/ron/Projects/sails/ap/node_modules/async/lib/async.js:615:34
    at /Users/ron/Projects/sails/ap/node_modules/async/lib/async.js:157:25
    at /Users/ron/Projects/sails/ap/api/models/User.js:48:21
```

My mocha config as follows:

```javascript
var Sails = require('sails');
var app;
before(function(done) {
  Sails.lift({
    connections: {
      mongo: {
        database: 'unit-test-db'
      }
    },
    models: {
      migrate: 'drop'
    },
    log: {
      level: 'debug'
    },
    hooks: {
      grunt: false,
      i18n: false
    }
  }, function(err, sails) {
    app = sails;
    done(err, app);
  });
});

after(function(done) {
   app.lower(function () {
      done(); 
    });
});
```

**Comment by YannBertrand** [comment]
+1

**Comment by jakutis** [comment]
+1

**Comment by ghadeerrahhal** [comment]
+1

My configuration is similar to yours. I get the same exact error when trying to validate a model. I get an error 3/5 times.

```node
var Q = require('q');

var validate = function (model, req) {
  return Q.promise(function(resolve) {
    model.validate(req.body, function(err) {
      resolve(err);
    });
  });
}

**Comment by ghadeerrahhal** [comment]
Upon further investigation, there is definitely some sort of race condition happening. To get around the issue, I set a timeout in my test and I no longer see a problem. This is definitely not ideal but helps work around the issue for the time being

```node
var Q = require('q');

var validate = function (model, req) {
  return Q.promise(function(resolve) {
    setTimeout(fn, 1000)

    function fn() {
      model.validate(req.body, function(err) {
        resolve(err);
      })
    };
  });
}

**Comment by sailsbot** [comment]
Thanks for posting, @ronsuez.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by akocmark** [comment]
same issue for running multiple concurrent inserts

**Comment by devinivy** [comment]
@akocmark what version of the adapter are you using?  Sounds like a bug!

**Comment by Josebaseba** [comment]
Same problem here:

```javascript
before(function(done) {
  Sails.lift({
    // configuration for testing purposes
    environment: 'test',
    bootstrap: testBootstrap,
    log: {level: process.env.log || 'error'}
  }, function(err, sails) {
    if(err) return done(err);
    if(sails.config.environment !== 'test') return done('Please, check the environment');
    // Load mockup data
    loadDb.start(function(err){
      //start tests
      if(err) return done(err);
      done(err, sails);
    })
  });
});

after(function(done) {
    sails.lower(done); // Here I found the same crash
});
```

But if I add a setTimeout it works:

```javascript
after(function(done) {
  setTimeout(function(){
    sails.lower(done);
  }, 200);
});
```

I've tried with sails-mongo 0.10 and 0.11, with sails 0.12 it's breaking but it works with sails 0.10.5.

**Comment by akocmark** [comment]
@devinivy it's 0.10.8 should I update to 0.11.x?

**Comment by devinivy** [comment]
Yes, I suggest upgrading!  Note that the mongodb driver has been upgrade to v2 in sails-mongo v0.11.

**Comment by daniellizik** [comment]
Same problem using sails 0.11.2 and sails-mongo 0.11.4. The weird thing is sometimes my tests will run and sometimes I get the error. Adding a timeout in the before or after callback doesn't do anything for me either.

    var Sails = require("sails");
    var sails;

    before(function(done) {
        Sails.lift({}, function(err, app) {
            sails = app;
            if (err) return done(err);
            done(err, sails);
        });
    });

    after(function(done) {
        sails.lower(done);
    });

    =======

    TypeError: Cannot read property 'collections' of undefined
    at Object.module.exports.adapter.update (/home/daniel/portal/node_modules/sails-mongo/lib/adapter.js:351:40)
    at module.exports.update (/home/daniel/portal/node_modules/sails/node_modules/waterline/lib/waterline/adapter/dql.js:241:13)
    at bound.updateRecords (/home/daniel/portal/node_modules/sails/node_modules/waterline/lib/waterline/query/dql/update.js:201:16)
    at /home/daniel/portal/node_modules/sails/node_modules/waterline/lib/waterline/query/dql/update.js:52:21

**Comment by craigvl** [comment]
Same issue for me and the timeout does not fix:
Sails 0.11.0
sails-mongo 0.11.4

![image](https://cloud.githubusercontent.com/assets/3496365/11977139/aed5e138-a9cb-11e5-9e11-84e533cc801d.png)

**Comment by particlebanana** [comment]
Can anyone create a repo I can clone and reproduce this? It looks like everyone is having issues in testing. From the stack traces shown here it looks like you are trying to use a connection that has not been defined. If thats the case I need to add a better error message in there.

I need a bit of clarification on what exactly the issue is though now that most of these line numbers are outdated.

**Comment by Josebaseba** [comment]
Hey @particlebanana, I think that I found the reason (at least in my case). What was happening in my tests was that when I was running the `sails.lower()` command, waterline was trying to keep doing some stuff. I'm going to try to create a similar example of what was happening in my app:

```javascript

// Controller

getSomeInfo: function(req, res){
  Info.findOne(req.param('id')).exec(function(err, info){
      if(err) return res.serverError(err);
      if(!info) return res.notFound();
      info.requestedTimes += 1;
      info.save(function(err){
           if(err) return sails.log.error('Error adding +1 to info', err);
      });
      // Don't wait to the counter++ and send the data
      return res.json(info);
  });
}

// Test

it('should return 200 with exists true', function(done){
      request(sails.hooks.http.app)
        .get('/info/' + id)
        .expect(200)
        .end(function(err, res){
          if(err)return done(err);
          res.body.should.have.property('something').equal(true);
          done();
        });
    });

```

Then `sails.lower()` get into it and it turn off waterline, but `info.save(function(err){})` was trying to save the new data when waterline "turned off", so the error appear in the tests.

The weird thing is that with sails v0.10.5 the error doesn't appear.

Hope it helps!

**Comment by particlebanana** [comment]
@Josebaseba thanks for the clarification. I wonder if it was the same for others since it sounds like they experienced it in a test environment.

Going to close this for now.

**Comment by crobinson42** [comment]
+1 I'm experiencing this in all environments when I lower my app.

`TypeError: Cannot read property 'collections' of undefined`

using sails v0.12

**Comment by wlemahieu** [comment]
+1
Experiencing this in sails v0.11.2 when pulling in any changes whatsoever to any services & models. I haven't tested controllers. It's on a highly active website with 250k+ user requests a day so it makes sense that it happens during the time Waterline/MongoDB are still doing stuff. 

A workaround is using a separate DB than the default 'sails' db defined in connections.js if possible. It's more work but it does prevent that problem if you're testing.

**Comment by particlebanana** [comment]
@crobinson42 @wlemahieu are you lowering Sails programmatically and seeing this?

**Comment by crobinson42** [comment]
Ctrl+c

**Comment by particlebanana** [comment]
@crobinson42 can you give me details to reproduce or setup a repo that does it? I'm not seeing this on any apps or in any tests so any help would be appreciated.

**Comment by wlemahieu** [comment]
@particlebanana I believe it's only happening when a read / write occurs on the same collection for the model file I'm editing. So if the app is constantly writing to CollectionA and I make a change to the CollectionA.js model file, simply making that file-save causes the error above. Having a separate DB for production solved the whole mess and is better practice though.

I never see this problem when using a test app that doesn't have a lot of read/writes going on for the collection we're editing, so the tests need to emulate heavy traffic.

**Comment by particlebanana** [comment]
@wlemahieu not sure I follow. Do you have something restarting your app when you save a file? Essentially this errors means something has already released the connection and the server no longer knows about it.

**Comment by wlemahieu** [comment]
@particlebanana I have production running on port 1338 and development running on port 1337. Both SailsJS apps were using the 'sails' mongoDB, however. So they were tied together in that aspect. When I would make an update to CollectionA.js, which obviously is for 'CollectionA' collection in mongoDB, I would receive that error.

This happened without lifting or lowering. It happens as soon as I make a save to that model file on the development folder structure. I solved it by using a prod/swap/dev DB system to avoid bringing my app down. Although Forever would bring it up right away, having simple file-saves cause this disruption is weird.

**Comment by crobinson42** [comment]
Ahh, I see a correlation with Wesley ^^^. I also run on 1337 & 1338 i.e.: production server to support zero down time update/rolls.

**Comment by crobinson42** [comment]
@particlebanana 
> @Josebaseba thanks for the clarification. I wonder if it was the same for others since it sounds like they experienced it in a test environment.

This happens in production and staging servers as well as local dev for me. #398 

It's related to my socket logic that attempts to write to my DB connection `afterDisconnect` which the socket is consequently disconnected because Sails is "lowering". Every time I deploy/roll my prod server I get this because sockets are connected.

### Solution Proposal
Offer a global (ie: `sails.config.sails_is_lowering`) so for those that experience this issue in my case we can put simple logic in place to test if the app is "lowering" programmatically.
